mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -1284,42 +1231,99 @@ private void executeBatchRemove(Map<DBCollection, BatchInsertionTask> inserts, R
 		}
 	}
 
-	private void executeBatchUpdate(Map<DBCollection, BatchInsertionTask> inserts, InsertOrUpdateTupleOperation tupleOperation) {
-		EntityKey entityKey = tupleOperation.getEntityKey();
-		Tuple tuple = tupleOperation.getTuple();
-		MongoDBTupleSnapshot snapshot = (MongoDBTupleSnapshot) tupleOperation.getTuple().getSnapshot();
-		WriteConcern writeConcern = getWriteConcern( tupleOperation.getTupleContext() );
-
-		if ( INSERT == snapshot.getSnapshotType() ) {
-			prepareForInsert( inserts, snapshot, entityKey, tuple, writeConcern );
-		}
-		else {
-			// Object already exists in the db or has invalid fields:
-			insertOrUpdateTuple( entityKey, tuple, tupleOperation.getTupleContext() );
-		}
-	}
-
-	private void executeBatchUpdateAssociation(Map<DBCollection, BatchInsertionTask> inserts, InsertOrUpdateAssociationOperation updateOp) {
-		AssociationKey associationKey = updateOp.getAssociationKey();
-		if ( isEmbeddedAssociation( associationKey ) ) {
-			DBCollection collection = getCollection( associationKey.getEntityKey() );
-			BatchInsertionTask batchInserts = inserts.get( collection );
-			if ( batchInserts != null && batchInserts.containsKey( associationKey.getEntityKey() ) ) {
-				// The owner of the association is in the insertion queue,
-				// we are going to update it with the collection of elements
-				WriteConcern writeConcern = getWriteConcern( updateOp.getContext() );
-				BatchInsertionTask insertTask = getOrCreateBatchInsertionTask( inserts, associationKey.getEntityKey().getMetadata(), collection, writeConcern );
-				DBObject documentForInsertion = insertTask.get( associationKey.getEntityKey() );
-				Object embeddedElements = getAssociationRows( updateOp.getAssociation(), updateOp.getAssociationKey(), updateOp.getContext() );
+	private void executeBatchUpdate(Map<DBCollection, BatchInsertionTask> inserts, List<Tuple> insertTuples,
+			GroupedChangesToEntityOperation groupedOperation) {
+		EntityKey entityKey = groupedOperation.getEntityKey();
+		DBCollection collection = getCollection( entityKey );
+		DBObject insertStatement = null;
+		BasicDBObject updateStatement = new BasicDBObject();
+		WriteConcern writeConcern = null;
+
+		for (Operation operation : groupedOperation.getOperations()) {
+			if ( operation instanceof InsertOrUpdateTupleOperation ) {
```

Comment:
Yes, Gunnar also expressed concerns about the fact that the method was bigger than what we usually recommend. I must admit that it's done like this on purpose even if it's not pretty: it's far easier to debug and see how it interacts like that (and trust me, I spent a lot of time debugging it ;)).

I suppose it's a general comment: for the AbstractGroupingEntityDialect implementations, it might change if we can move some code to the base class.